extension-androidauto/src/main/java/com/mapbox/maps/extension/androidauto/CarMapLifecycleObserver.kt

```diff
@@ -0,0 +1,128 @@
+ * our own [MapboxCarMapObserver]
+ */
+internal class CarMapLifecycleObserver internal constructor(
+  private val carContext: CarContext,
+  private val carMapSurfaceOwner: CarMapSurfaceOwner,
+  private val mapInitOptions: MapInitOptions
+) : DefaultLifecycleObserver, SurfaceCallback {
+
+  private var mapStyleUri: String = mapInitOptions.styleUri ?: Style.MAPBOX_STREETS
+
+  private val logMapError = object : OnMapLoadErrorListener {
+    override fun onMapLoadError(eventData: MapLoadingErrorEventData) {
+      val errorData = "${eventData.type} ${eventData.message}"
+      Logger.e(TAG, "updateMapStyle onMapLoadError $errorData")
+    }
+  }
+
+  /** Screen lifecycle events */
+
+  override fun onCreate(owner: LifecycleOwner) {
+    Logger.i(TAG, "onCreate request surface")
+    carContext.getCarService(AppManager::class.java)
+      .setSurfaceCallback(this)
+  }
+
+  /** Surface lifecycle events */
+
+  override fun onSurfaceAvailable(surfaceContainer: SurfaceContainer) {
+    Logger.i(TAG, "onSurfaceAvailable $surfaceContainer")
+    surfaceContainer.surface?.let { surface ->
+      val mapSurface = MapSurfaceProvider.create(
+        carContext,
+        surface,
+        mapInitOptions
+      )
+      mapSurface.onStart()
+      mapSurface.surfaceCreated()
+      mapSurface.getMapboxMap().loadStyleUri(
+        mapStyleUri,
+        onStyleLoaded = { style ->
+          Logger.i(TAG, "onSurfaceAvailable onStyleLoaded")
+          mapSurface.surfaceChanged(surfaceContainer.width, surfaceContainer.height)
+          val carMapSurface = MapboxCarMapSurface(carContext, mapSurface, surfaceContainer, style)
+          carMapSurfaceOwner.surfaceAvailable(carMapSurface)
+        },
+        onMapLoadErrorListener = logMapError
+      )
+    }
+  }
+
+  override fun onVisibleAreaChanged(visibleArea: Rect) {
+    Logger.i(TAG, "onVisibleAreaChanged visibleArea:$visibleArea")
+    carMapSurfaceOwner.surfaceVisibleAreaChanged(visibleArea)
+  }
+
+  override fun onStableAreaChanged(stableArea: Rect) {
+    // Have not found a need for this.
+    // logAndroidAuto("CarMapSurfaceLifecycle Stable area changed stable:$stableArea")
+  }
+
+  override fun onScroll(distanceX: Float, distanceY: Float) {
+    Logger.i(TAG, "onScroll $distanceX, $distanceY")
+    carMapSurfaceOwner.scroll(distanceX, distanceY)
+  }
+
+  override fun onFling(velocityX: Float, velocityY: Float) {
+    Logger.i(TAG, "onFling $velocityX, $velocityY")
+    carMapSurfaceOwner.fling(velocityX, velocityY)
+  }
+
+  override fun onScale(focusX: Float, focusY: Float, scaleFactor: Float) {
+    Logger.i(TAG, "onScroll $focusX, $focusY, $scaleFactor")
+    carMapSurfaceOwner.scale(focusX, focusY, scaleFactor)
+  }
+
+  override fun onSurfaceDestroyed(surfaceContainer: SurfaceContainer) {
+    Logger.i(TAG, "onSurfaceDestroyed")
+    carMapSurfaceOwner.surfaceDestroyed()
+  }
+
+  /** Map modifiers */
+
+  fun updateMapStyle(styleUri: String) {
```

Comment:
I would suggest to remove these style related methods from the extension, as user might want to do some complex runtime styling and set the style themselves. And by default, user could configure the default style using mapInitOptions and the style will load automatically already.

Reply:
This is a convenience function to make it easy to switch between night/day modes. And then all your downstream `MapboxCarMapObserver`s will `detach` and `attach`.

``` kotlin
internal class CarMapSurfaceOwner
--snip--
  fun surfaceAvailable(mapboxCarMapSurface: MapboxCarMapSurface) {
--snip--
    oldCarMapSurface?.let { carMapObservers.forEach { it.detached(oldCarMapSurface) } }
    carMapObservers.forEach { it.loaded(mapboxCarMapSurface) }
```

This makes it really easy for a developer to handle night and day style switching, and all of the `MapboxCarMapObserver`s are notified.

``` kotlin
  override fun onCarConfigurationChanged(newConfiguration: Configuration) {
    val mapStyle = if (carContext.isDarkMode) Style.TRAFFIC_NIGHT else Style.TRAFFIC_DAY
    mapboxCarMap.updateMapStyle(mapStyle())
  }
```

So this is different from `MapInitOptions` which are part of the map construction. 
``` kotlin
internal class CarMapLifecycleObserver internal constructor(
  private val carContext: CarContext,
  private val carMapSurfaceOwner: CarMapSurfaceOwner,
  private val mapInitOptions: MapInitOptions
) : DefaultLifecycleObserver, SurfaceCallback {
```

Reply:
It seems a bit weird that LifecycleObserver is responsible for managing styles. 
Would be nice to have some separate class for that?

Reply:
> It seems a bit weird that LifecycleObserver is responsible for managing styles.
Would be nice to have some separate class for that?

One of the difficult problems that we needed to solve in order to make a reliable solution was, handling "Surface + Mapbox Style + Lifecycle" race conditions.  https://github.com/mapbox/1tap-android/issues/1008

So this class is merging these events and projecting them onto a single `MapboxCarMapObserver` interface. 

@yunikkk Can you be a little more specific about what is weird about that though, are you suggesting a refactor?

Reply:
> One of the difficult problems that we needed to solve in order to make a reliable solution was, handling "Surface + Mapbox Style + Lifecycle" race conditions. [mapbox/1tap-android#1008](https://github.com/mapbox/1tap-android/issues/1008)
> 
> So this class is merging these events and projecting them onto a single `MapboxCarMapObserver` interface.

I see. Will look closer into the tickets, not sure what's the good way to avoid races correctly. From the first glance it seems maps SDK may have some issues with style callbacks.

> Can you be a little more specific about what is weird about that though, are you suggesting a refactor?

So it seems weird that `LifecycleObserver` is also responsible for Style management. From the name itself I've expected it will listen for lifecycle / surface callbacks and forward it somewhere else (CarMapSurface for example). 
But it also manages style, contains some internal state (mapStyleUri). 
Firstly - rest of style setting options would also need to be implemented (StyleContract, json string). 
Also - it doesn't feel like a class with single responsibility. Moreover, surface create logic is handled in this observer, while destroy - in [CarSurfaceOwner](https://github.com/mapbox/mapbox-maps-android/pull/1004/files#diff-eaac5e9b61e5347cff7faf58f318352942c01165018b001283c016193bcd6a62R63), e.g.
```
    detachSurface?.mapSurface?.onStop()
    detachSurface?.mapSurface?.surfaceDestroyed()
    detachSurface?.mapSurface?.onDestroy()
```
I would expect CarSurfaceOwner to perform both since it `owns` the surface, while LifecycleObserver just listens to events and forwards them accordingly.

Reply:
@yunikkk would you like to propose some code in 1tap?

End of the day, this solution is copy/pasted from there where we are testing solutions in real cars. I wouldn't want to make a large refactor here, because we also need to make sure it works

Reply:
Added the `@MapboxExperimental` annotation. We can continue to iterate here. I am in agreement that, there are many ways to update the map style. having this function, and an interface that depends on styles changing this way, would cause issue.

But we also need a new solution that works. Linking to this other thread https://github.com/mapbox/mapbox-maps-android/pull/1004#discussion_r789581616. We essentially need a new proposal for observing map style changes

Reply:
Seems would be nice to address the questions before rolling out the extension to the general public. I'll cut separate branch from yours to play with code a bit and try to suggest smth here.
Don't think it'll be big refactor for 1tap anyway.

Reply:
Resolving this discussion and moving here https://github.com/mapbox/1tap-android/issues/1594